interpreter/src/cursor.rs

```diff
@@ -0,0 +1,3003 @@
+                push(&expression.expression);
+                false
+            }
+            Statement::Expression(_) if step == 1 => {
+                self.values.pop();
+                true
+            }
+            Statement::Iteration(iteration) if step == 0 => {
+                assert!(!iteration.inclusive);
+                push(&iteration.stop);
+                push(&iteration.start);
+                false
+            }
+            Statement::Iteration(iteration) => {
+                // Currently there actually isn't a syntax in Leo for inclusive ranges.
+                let stop = self.pop_value()?;
+                let start = self.pop_value()?;
+                if start.eq(&stop) {
+                    true
+                } else {
+                    let new_start = start.inc_unchecked();
+                    self.contexts.set(iteration.variable.name, start);
+                    self.frames.push(Frame {
+                        step: 0,
+                        element: Element::Block { block: &iteration.block, function_body: false },
+                        user_initiated: false,
+                    });
+                    self.values.push(new_start);
+                    self.values.push(stop);
+                    false
+                }
+            }
+            Statement::Return(return_) if step == 0 => {
+                push(&return_.expression);
+                false
+            }
+            Statement::Return(_) if step == 1 => loop {
+                let last_frame = self.frames.last().expect("a frame should be present");
+                match last_frame.element {
+                    Element::Expression(Expression::Call(_)) | Element::DelayedCall(_) => {
+                        if self.contexts.is_async() {
+                            // Get rid of the Unit we previously pushed, and replace it with a Future.
+                            self.values.pop();
+                            self.values.push(Value::Future(self.contexts.get_future()));
+                        }
+                        self.contexts.pop();
+                        return Ok(true);
+                    }
+                    _ => {
+                        self.frames.pop();
+                    }
+                }
+            },
+            _ => unreachable!(),
+        };
+
+        if done {
+            assert_eq!(len, self.frames.len());
+            self.frames.pop();
+        } else {
+            self.frames[len - 1].step += 1;
+        }
+
+        Ok(done)
+    }
+
+    fn step_expression(&mut self, expression: &'a Expression, step: usize) -> Result<bool> {
+        let len = self.frames.len();
+
+        macro_rules! push {
+            () => {
+                |expression| {
+                    self.frames.push(Frame { element: Element::Expression(expression), step: 0, user_initiated: false })
+                }
+            };
+        }
+
+        if let Some(value) = match expression {
```

Comment:
Organizational nit. Consider splitting the functional components (expression evaluation) and control flow of the REPL loop for easier maintenance.

Reply:
Ah that's my bad on the lack of clarity. I was thinking splitting them across different files.